invidious: unstable-2021-11-08 -> unstable-2022-03-16

schnusch opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Add update script.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sbruder committed:
invidious: unstable-2021-11-08 -> unstable-2021-11-13

This update disables QUIC by default which fixes Invidious not loading
anything except for the home page due to YouTube no longer accepting
HTTP/3 (Upstream Issue:
https://github.com/iv-org/invidious/issues/2577).

It therefore uses Crystal’s internal HTTP client, which failed because
the statically linked boringssl (required by lsquic) overrides OpenSSL’s
CA certificate file location. This is fixed by applying the same patch
to boringssl that is applied to openssl for using the correct CA
certificate file.

schnusch committed:
invidious: add update script

based on mpvScripts.sponsorblock's update script

sbruder commented on pkgs/servers/invidious/lsquic.nix:
```diff
@@ -2,23 +2,23 @@
 stdenv.mkDerivation rec {
   pname = "lsquic";
-  version = "2.18.1";
+  version = "3.0.3";
```
I didn’t test this yet, and I don’t know if I’ll have the time to test it in the next days, but AFAIK `lsquic.cr` always requires the lsquic version to match the tag of its version (see [README](https://github.com/iv-org/lsquic.cr#lsquic)). lsquic also requires a specific boringssl version, also specified in [its README](https://github.com/litespeedtech/lsquic#building-boringssl).

schnusch replied:
You are probably right, so the update script should pick the lsquic version from *shard.lock*.

sbruder commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,92 @@
+invidious_git='/var/tmp/invidious.git'
+
+set -euo pipefail
+
```
I think this should explicitly change the directory to the root of the package, because otherwise `nix-shell maintainers/scripts/update.nix --argstr package invidious` fails.

```suggestion
cd "$(dirname "${BASH_SOURCE[0]}")"

```

sbruder reviewed (changes requested):
Apart from the one thing I noted below, this looks nice and works, thank you!

sbruder reviewed (approved):
Thanks for addressing the feedback, this looks good.

schnusch committed:
invidious: unstable-2021-11-13 -> unstable-2022-02-25

schnusch commented:
With <https://github.com/iv-org/invidious/pull/2685> `substituteInPlace src/invidious/helpers/helpers.cr ...` needs fixing and other stuffs breaks during build as well.

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,99 @@
+#!nix-shell -i bash -p common-updater-scripts curl crystal crystal2nix jq git gnused nix nix-prefetch-git nix-update pkg-config
+git_url='https://github.com/iv-org/invidious.git'
+git_branch='master'
+git_dir='/var/tmp/invidious.git'
```
```suggestion
tmp=$(mktemp -d)
trap 'rm -rf "$tmp"' exit
git_dir="$tmp/invidious.git"
```

This way there's a clean new temporary directory every time. Or is it intentional that you're reusing this temporary directory?

schnusch replied:
I think [/var/tmp](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s15.html) is made for this.

Infinisil replied:
Interesting, I initially thought this was like how `/var/run` is symlinked to `/run` on NixOS, the same doesn't hold for `/var/tmp` and `/tmp` though. I'm thinking that `/var/tmp` is about the system-wide equivalent to the user-local `~/.cache`.

I think I'd probably still prefer the `XDG_CACHE_HOME` with the `~/.cache` fallback since it's configurable and a fairly popular [specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), but I think `/var/tmp` is fine too.

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,99 @@
+    # make lsquic available as an attribute in nixpkgs, so we can use nix-update
+    tmp=$(mktemp --suffix='.nix')
+    trap 'rm -f "$tmp"' EXIT
+    cat << eof > "$tmp"
+args:
+let
+  pkgs = import ${PWD%/*/*/*} args;
+in
+  pkgs // { lsquic = pkgs.callPackage $PWD/lsquic.nix {}; }
+eof
+
+    lsquic_old_version=$(nix-instantiate --eval --strict --json -A lsquic.version --expr "import $tmp {}" | jq -r)
```
Is this done like this because `lsquic` is not exported from `pkgs.*`? If so, I suggest setting `passthru.lsquic` in the invidious derivation, so that you don't have to hack around this and can use `pkgs.invidious.lsquic.version` instead.

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,99 @@
+        if git diff-index --quiet HEAD -- 'lsquic.nix'; then
+            info "lsquic is up-to-date."
+        else
+            boringssl_new_version=$(curl -LSsf "https://github.com/litespeedtech/lsquic/raw/v${lsquic_new_version}/README.md" \
+                | grep -Pom1 '(?<=^git checkout ).*')
```
Oh my, this is really hacky, but it [does appear](https://github.com/litespeedtech/lsquic/pull/346/files) to be how upstream updates their version. We should definitely report this upstream though, that's not good and should be fixed

Infinisil commented on pkgs/servers/invidious/videojs.sh:
```diff
@@ -0,0 +1,7 @@
+
+unpackPhase
+cd source
+crystal run scripts/fetch-player-dependencies.cr -- --minified
```
I think a comment mentioning why this is pure would be nice

infinisil reviewed (changes requested):
In addition to the below comments, I feel like the update script would be much less hacky if the versions and hashes were stored in JSON files, which are then imported from Nix, instead of Nix directly.

schnusch committed:
invidious: move versions, revs and sha256s to JSON file

This way updating these values is much less hacky. A downside is that tools
like nix-update will no longer be able to update these, but this should not
be too important for invidious.

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -16,8 +18,12 @@ info() {
+    jq --arg x "$2" "$1 = \$x" < 'versions.json' | sponge 'versions.json'
+}
+
+old_rev=$(nix-instantiate --eval --strict --json -A "$pkg.src.rev" '../../..' | jq -r)
+old_version=$(nix-instantiate --eval --strict --json -A "$pkg.version" '../../..' | jq -r)
```
You can read the `versions.json` file here directly

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,89 @@
+    info "Updating shards.nix..."
+    crystal2nix -- "$git_dir/shard.lock"  # argv's index seems broken
+
+    lsquic_old_version=$(nix-instantiate --eval --strict --json -A "${pkg}.lsquic.version" '../../..' | jq -r)
```
Same here

Infinisil commented on pkgs/servers/invidious/update.sh:
```diff
@@ -0,0 +1,89 @@
+
+    lsquic_old_version=$(nix-instantiate --eval --strict --json -A "${pkg}.lsquic.version" '../../..' | jq -r)
+    lsquic_new_version=$(nix eval --raw -f 'shards.nix' lsquic.rev \
+        | sed -e 's/^v//' -e 's/-[0-9]*$//')
```
```suggestion
        | sed -e 's/-[0-9]*$//')
```

Then you don't need to re-add the `v` prefix in `lsquic.nix`. This should also have a comment explaining why this is needed, which is because [lsquic.cr](https://github.com/iv-org/lsquic.cr)'s versions track [lsquic](https://github.com/litespeedtech/lsquic)'s (see https://github.com/iv-org/lsquic.cr/blob/master/README.md?plain=1#L5), but if there are multiple lsquic.cr releases for the same lsquic one, they seem to add a `-<number>` suffix.

schnusch replied:
We must still strip the `v` when we put it in the *versions.json*. I would rather just strip it at the start to keep the variable "clean".

schnusch committed:
invidious: unstable-2022-02-25 -> unstable-2022-03-16

isker commented:
Thanks for maintaining this branch for so long.  I hope it gets merged soon 🌞.

Infinisil commented:
Just tested this by cherry-picking this PR, enabling `services.invidious.enable`, going to `localhost:3000` and doing any search. Unfortunately this runs into this issue, which makes it essentially unusable: https://github.com/iv-org/invidious/issues/2753

Title: `SSL_connect: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED (OpenSSL::SSL::Error)`
Date: `2022-03-12T02:41:05Z`
Route: `/search?q=aoeu`
Version: `2022.03.07-440549f @ master`

<details>
<summary>Backtrace</summary>
<p>
   
```
SSL_connect: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED (OpenSSL::SSL::Error)
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from /nix/store/wpiwbvbcia4w9qlmdcbdvwkiqavxcg7l-invidious-unstable-2022-03-07/bin/invidious in &#39;??&#39;
  from ???
```
</p>
</details>

This doesn't happen with the old version without this PR, so this is a regression :/

sbruder commented:
The reason for this issue is that the statically linked boringssl (required by lsquic) overrides OpenSSL’s CA certificate file location.

It can be fixed either by patching boringssl (like in https://github.com/NixOS/nixpkgs/pull/145882) or by setting the `SSL_CERT_FILE` environment variable (like @squalus pointed out in https://github.com/NixOS/nixpkgs/pull/145882#issuecomment-972104021).

<details><summary>Patching boringssl</summary>

```diff
diff --git a/pkgs/servers/invidious/lsquic.nix b/pkgs/servers/invidious/lsquic.nix
index e44e41cee09..50f6d186705 100644
--- a/pkgs/servers/invidious/lsquic.nix
+++ b/pkgs/servers/invidious/lsquic.nix
@@ -9,6 +9,11 @@ let
       url = "https://boringssl.googlesource.com/boringssl";
       inherit (versions.boringssl) rev sha256;
     };
+
+    patches = (old.patches or [ ]) ++ [
+      # Use /etc/ssl/certs/ca-certificates.crt instead of /etc/ssl/cert.pem
+      ./use-etc-ssl-certs.patch
+    ];
   });
 in
 stdenv.mkDerivation rec {
diff --git a/pkgs/servers/invidious/use-etc-ssl-certs.patch b/pkgs/servers/invidious/use-etc-ssl-certs.patch
new file mode 100644
index 00000000000..ea62378008d
--- /dev/null
+++ b/pkgs/servers/invidious/use-etc-ssl-certs.patch
@@ -0,0 +1,13 @@
+diff --git a/crypto/x509/x509_def.c b/crypto/x509/x509_def.c
+index d2bc3e5c1..329580075 100644
+--- a/crypto/x509/x509_def.c
++++ b/crypto/x509/x509_def.c
+@@ -67,7 +67,7 @@
+
+ #define X509_CERT_AREA          OPENSSLDIR
+ #define X509_CERT_DIR           OPENSSLDIR "/certs"
+-#define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
++#define X509_CERT_FILE          "/etc/ssl/certs/ca-certificates.crt"
+ #define X509_PRIVATE_DIR        OPENSSLDIR "/private"
+ #define X509_CERT_DIR_EVP        "SSL_CERT_DIR"
+ #define X509_CERT_FILE_EVP       "SSL_CERT_FILE"
```
</details>

<details><summary>Setting the SSL_CERT_FILE environment variable with a wrapper</summary>

```diff
diff --git a/pkgs/servers/invidious/default.nix b/pkgs/servers/invidious/default.nix
index 14eed9a3be7..cfe423236f5 100644
--- a/pkgs/servers/invidious/default.nix
+++ b/pkgs/servers/invidious/default.nix
@@ -56,7 +56,7 @@ crystal.buildCrystalPackage rec {
           --replace 'Process.run(%(rsvg-convert' 'Process.run(%(${lib.getBin librsvg}/bin/rsvg-convert'
     '';

-  nativeBuildInputs = [ pkg-config shards ];
+  nativeBuildInputs = [ makeWrapper pkg-config shards ];
   buildInputs = [ libxml2 openssl sqlite ];

   format = "crystal";
@@ -77,6 +77,9 @@ crystal.buildCrystalPackage rec {
   '';

   postInstall = ''
+    wrapProgram $out/bin/invidious \
+        --set SSL_CERT_FILE "/etc/ssl/certs/ca-bundle.crt" \
+
     mkdir -p $out/share/invidious/config

     # Copy static parts
```
</details>

Applying any of those patches fixed the problem for me.

sbruder commented on pkgs/servers/invidious/default.nix:
```diff
@@ -8,17 +8,16 @@ let
   #  * shards.nix (by running `crystal2nix` in invidious’ source tree)
   #  * If the lsquic.cr dependency changed: lsquic in lsquic.nix (version, sha256)
   #  * If the lsquic version changed: boringssl' in lsquic.nix (version, sha256)
```
I think this should be changed to reflect that the versions/hashes are now stored in a separate file (and no longer in `lsquic.nix` or `default.nix`).

schnusch committed:
invidious: document update process

Infinisil commented:
The NixOS module should certainly be fixed before this can be merged, the `NIX_SSL_FILE` approach sounds good

schnusch commented:
sbruder's 970d249d39ccac356c705e6e4db6318ffe93dd42 fixes that.

schnusch commented:
sbruder's 970d249d39ccac356c705e6e4db6318ffe93dd42 fixes that.

infinisil reviewed (approved):
I think this is looking good! Can you squash down the commits so we only have a single update commit (that matches the PR title), and separate commits for the other changes?

schnusch commented:
I will squash them before the refactor in c3bae5b1191ddfdc036407884d603cff5cc6717f, otherwise it will get ugly.

Infinisil commented:
`aarch64-darwin` build fails in CI. Is that expected? If so the platform should be marked as not supported somewhere

schnusch commented:
The error `Failed to spawn codesign_allocate: No such file or directory` seems fixable, but I don't know the first thing about darwin, so I will mark it as broken.

schnusch committed:
invidious: mark as broken on aarch64-darwin

Build fails with
    Failed to spawn codesign_allocate: No such file or directory

Infinisil commented:
Looking good, thanks!
